Add simplified quantized tensor subclass from torchao

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134018

Summary:
This is an simplified quantized tensor subclass, we can use it for testing
subclass + compile + many other features

Test Plan:
python torch/testing/_internal/compilable_quantized_tensor.py

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Add simplified quantized tensor subclass from torchao

Summary:
This is an simplified quantized tensor subclass, we can use it for testing
subclass + compile + many other features

Test Plan:
python torch/testing/_internal/compilable_quantized_tensor.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add simplified quantized tensor subclass from torchao"


Summary:
This is an simplified quantized tensor subclass, we can use it for testing
subclass + compile + many other features

Test Plan:
python torch/testing/_internal/compilable_quantized_tensor.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,401 @@
+
+
+########
+# Test #
```
Note that test should be in a proper test file, not inline here.
But will let @bdhirsh comment here on how much he wants to see on this PR vs follow ups.

jerryzh168 replied:
sure, please let me know where should I put it

jerryzh168 committed:
Update on "Add simplified quantized tensor subclass from torchao"


Summary:
This is an simplified quantized tensor subclass, we can use it for testing
subclass + compile + many other features

Test Plan:
python torch/testing/_internal/compilable_quantized_tensor.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+######################################################
+
+
+class PlainMyDTypeLayout(MyDTypeLayout):
```
I'd prefer to keep this class simple-and-modifiable over flexible-and-complicated.

In that vein, can we remove all of the multiple layers of inheritance, and just have a single subclass that inherits from `torch.Tensor` and exposes all of its logic?

bdhirsh replied:
In a similar vein, I was expecting a more specialized quantization example here. What flavor of quantization is this subclass doing? (what's the hardcoded dtype, is it hardcoded to row-wise quantization, etc).

jerryzh168 replied:
multiple layers of inheritance is actually what broke tensor_subclass + torch.compile for a lot of pytorch versions in the past actually, do we really want to remove it?

jerryzh168 replied:
sure I can update the name to be more specialized

bdhirsh replied:
yeah I think I would prefer to remove it. IIRC, most of the "bugs" we saw with multiple inheritance were actaully bugs in the sublcass impl, as opposed to bugs in the compiler that we needed to fix to get multiple inheritance working

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+it shows
+    * the basic structure of a new dtype tensor subclass (__new__, __init__, __tensor_flatten__, __tensor_unflatten__)
+    * two types of dispatch that people can overwrite (__torch_function__, __torch_dispatch__)
+    * how to abstract away packing format with layout
```
hmm I'm not too sure what this means

jerryzh168 replied:
oh please take a look at https://github.com/pytorch/ao/issues/391 I can link it here as well

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+    ):
+        kwargs = {}
+        kwargs["device"] = int_data.device
+        kwargs["layout"] = kwargs.get("layout") if kwargs.get("layout", False) else int_data.layout  # type: ignore[assignment]
```
nit: `kwargs` is guaranteed not to contain layout here, so you can just always do `int_data.layout`?

Can you help me understand the purpose of the layout here? My feeling is this is less necessary if we hardcode a specific quantization layout in this subclass

jerryzh168 replied:
yeah it's not strictly needed I think, we actually override the layout concept in native Tensor for something else, see `Pakcing/Layout` section in https://github.com/pytorch/ao/issues/391

bdhirsh replied:
got it, my preference would be to remove / hardcode one

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+    def __new__(  # type: ignore[no-untyped-def]
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
```
some comments on the required shapes for `int_data` vs `scale` would be good here. E.g. is it something like this?

`int_data: [M, N]`
`scale: [M]` (if this is row-wise scaling)

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+        """We can define a dequantize method to convert the quantized tensor to a floating point tensor"""
+        if output_dtype is None:
+            output_dtype = torch.get_default_dtype()
+        int_data, scale = self.layout_tensor.get_plain()
```
is this subclass meant to be used with training / fine-tuning?

I would definitely prefer that the example subclass we have be trainable and work with autograd (in both eager compile), so we can really advertise that it works in all of the composability tests we care about.

(My suspicion is because `self.layout_tensor.get_plain()` doesn't really seem differentiable)

jerryzh168 replied:
not yet, we can add this later, when @andrewor14 added a training example in https://github.com/pytorch/ao/tree/main/tutorials/developer_api_guide

bdhirsh replied:
I would like for the subclass that we put in core to start out with training support :)

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+    def __torch_dispatch__(cls, func, types, args, kwargs):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if func is aten.detach.default:
```
tbh if we want this subclass to be "best practices" that we are ok with people copy-pasting, I would prefer that we follow the registration pattern that float8 uses: https://github.com/pytorch/ao/blob/main/torchao/float8/float8_tensor.py#L359

Same goes for the NotImplemented stuff here: https://github.com/pytorch/ao/blob/main/torchao/float8/float8_tensor.py#L355

jerryzh168 replied:
we actually have some helper functions in torchao: https://github.com/pytorch/ao/blob/f5703b07acc683653556d04ef970709ba47dba10/torchao/utils.py#L330-L389

so this is not a best-practices example for people to copy pasting actually, the best practices example will be here: https://github.com/pytorch/ao/blob/main/tutorials/developer_api_guide/my_dtype_tensor_subclass.py, also we plan to add a few more things like training, tensor parallelism, maybe FSDP there as well

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+
+@implements(aten.detach.default)
+def _(func, types, args, kwargs):  # type: ignore[no-untyped-def]
+    # `return_and_correct_aliasing` should be used by wrapper tensor ``__torch_dispatch__`` subclasses that would like to
```
fwiw, we want to drop this requirement as part of https://docs.google.com/document/d/1vpbGMsbdhPzFpXi8GJxzkzAGufs0slrIqKqfwDNwKj4/edit#heading=h.76wykr4y5vap (cc @IvanKobzarev )

jerryzh168 replied:
you mean Ivan is working on this?

bdhirsh replied:
yep, it's on his list (we can keep the API in here for now though and remove it later)

bdhirsh commented on torch/testing/_internal/compilable_quantized_tensor.py:
```diff
@@ -0,0 +1,366 @@
+
+    def __repr__(self):  # type: ignore[no-untyped-def]
+        return (
+            f"{self.__class__.__name__}(data={self.dequantize()}, shape={self.shape}, "
```
nit: a more standard way we've been defining __repr__ is something like `return f"MySubclass(layout_tensor={repr(self.layout_tensor)}, scale={repr(self.scale)})"`

jerryzh168 replied:
yeah it's a choice of what we want to display, for quantized tensor typically the internal representation is not that meaningful for end users because they will just be packed tensors, that's why we typically just print the dequantized value that it's trying to represent
